README
GoVM
----



	1. VM Instructions



1.1 Dictionary


reg - Register
lit - Literal
add - Code address
dat - Data address

Note that except for hlt being represented by 0.0, no other instruction
has its value specified.
Also, `a` refers to the value of the first argument and `b` refers to the
value of the second argument, whenever applicable.



1.2 Instructions


THIS IS NOT UPDATED, SEE instructions.go FOR MISSING INSTRUCTIONS AND CHECK
THE CHANGES FILE FOR IMPLEMENTED CHANGES WHICH ADD THE NEW INSTRUCTIONS.


hlt
	Halts program execution
nop
	Does nothing.

val	[lit]	[reg]
	Sets b to be equal to a.
mov	[reg]	[reg]
	Sets b to be equal to a.
psh	[reg]
	Pushes a to the top of the stack.
pop	[reg]
	Pops the top of the stack into a.

add	[reg]	[reg]
	Adds a to b.
sub	[reg]	[reg]
	Subtracts a from b.
mul	[reg]	[reg]
	Multiplies a to b.
div	[reg]	[reg]
	Divides b by a.

inc	[reg]
	Increases a by 1.
dec	[reg]
	Decreases a by 1.
flr	[reg]
	Changes a to the highest integer value lower or equal than a.
cel	[reg]
	Changes a to the lowest integer value higher or equal than a.

shw	[reg]
	Prints the a to the standard output.
get	[reg]
	Sets a to whatever is entered into the standard input (if valid).
dsp	[dat]
	Prints the a-th string to the standard output.

cmp	[reg]	[reg]
	If a is less than b, flt is set to true. If a equals b, feq is set
	to true. Otherwise, fgt is set to true. When a flag is set to true
	the other flags are set to false.
cmz	[reg]
	Same behavior as cmp assuming there is a second argument where b
	equals zero.

jmp	[add]
	Jumps to the a-th instruction.
jgt	[add]
	Jumps to the a-th instruction if the fgt flag is set to true.
jeq	[add]
	Jumps to the a-th instruction if the feq flag is set to true.
jlt	[add]
	Jumps to the a-th instruction if the flt flag is set to true.

cll	[add]
	Jumps to the a-th instruction and saves current instruction index
	plus two (the next instruction) into the call stack.
ret
	Let n be the topmost value in the call stack. Jumps to the n-th
	instruction and pops the call stack.

	2. Syntax



2.1 Registers


All registers are written as %n where n is an *integer* referring to the
n-th register.



2.2 Comments


Any text after an instruction and its arguments is a valid comment.
Please mind that str pseudo-instructions consume everything so don't
comment on those lines.

Examples:

nop Hey jude.
add %0	%1 Don't say goodbye.



2.3 Jump statements

Label syntax is specified by PC006.

Jump syntax basically works as `j** <labelname>`.

Label redefinition is specified by PC002.
Undefined labels behavior is specified by PC014.

Also, jumps to labels that preceed no instruction is currently undefined behavior.

2.4 Constant strings

String syntax also has special pseudo-instructions. 

str	<name>	...

Basically sets <name> to represent the index a string made of every word after
it's name but separated by a space.
I don't know what happens if there's nothing but whitespace after <name>, since
I have not implemented this yet. PS: It's just newline.
To display a string, you use

dsp	<name>

so that you don't have to know by heart all of the string indexes.
The <name> used in a dsp instruction has to be already declared by a str.
If you don't it will be actually set as 0, so as long as you defined at
least one string, the first one will be defaulted to.

2.5 Multiple files


See PC004 in CHANGES for reference.
Example of this:
./govm compile examples/fibonacci/fib examples/fibonacci/main.gs examples/fibonacci/fib.gs
./govm examples/fibonacci/fib 20

2.6 Command Line Arguments


See PC008 in CHANGES for reference.


CHANGES
	1. Implemented

PC001:  2016-09-16
  Change label evaluation to only occur after all parsing instead of evaluating
  immediately in case the specific label has already been parsed.
PC002:  2016-09-16
  Make label redefinition into a parsing/compiler error.
PC003:  2016-09-16
  Change cpy to mov. mov makes more sense traditionally for this specific function
  and I want cpy freed for something in the future regarding memory access since
  it feels more appropriate for that.
PC005:  2016-09-16
  Since PC004 solves the possible problem of needing to separate regular 'jumping'
  labels from function labels, the def instruction will be deprecated in favor of
  just lab.
PC015:  2016-09-16
  Remove any restriction on code size and make everything dynamic. I can afford to
  just append at each new instruction in the parsing step.
PC014:  2016-09-16
  Make it so referencing an undefined label OR a label for the 0-th address is
  a compilation error and make sure the compilation error prints the line at fault.
PC016:  2016-09-16
  Make it so compilation errors are file aware.
PC017:  2016-09-16
  Make it so label evaluation happens with a function call after parsing files.
PC004:  2016-09-16
  Allow multiple files. The first file is where execution begins, and each subse-
  quent file acts as if appended to the first one. This allows for continuation of
  execution through files, but it's intended to allow for function labels defined
  in other files to be easily accessible with the current design.
PC013:  2016-09-16
  Make it so the nop instruction actually compiles into nothing.
PC006:  2016-09-16
  Deprecate lab label definition in favor of the "<name>:" syntax. In the switch
  for the first word in the line, the default case should be changed to check if
  the last byte of the word equals ':', if so, do the same thing "lab" does except
  instead of a second word, use every byte but the last one of the first word.
  Note that because of how parsing works, any instructions on the same line as the
  label will be treated as comments.
PC010:  2016-09-16
  Implement a set of mathematical functions as instructions.
  abs reg     Absolute value
  sqr reg     Square root
  sin reg     Sine
  cos reg     Cosine
  tan reg     Tangent
  exp reg     Exponential
  log reg     Natural logarithm
  gam reg     Gamma
  pow reg reg Sets b to the value of b^a
  pwa reg reg Sets b to the value of a^b
  These should be implemented as basic function calls to the math stdlib module.
PC019:  2016-09-16
  This PC specifies a binary format.
  The current way things work, global values are set during parsing so that they
  can be used by the runtime, namely the data stack. This data needs to be passed
  to the binary file.
  The encoding is little endian as opposed to big endian so that the files are smaller :)
  Here's the binary format:
  First 8 bytes: int64 header, currently equal to 10002.
  Next 8 bytes: size of the code.
  Variable size section containing all of the binary code.
  Next 8 bytes: int64 containing the size of the data stack.
  Variable size section containing all strings of the data stack, all ended
  by a '\n' byte.
PC018:  2016-09-16
  This PC specifies compilation and execution steps.
  Separate compilation and execution.
  For compilation into a binary bytecode file do:
  govm compile binary_name main.gs [other source files]
  For execution run:
  govm binary_name [arguments]
  This depends on PC019.
PC008:  2016-09-16
  Implement a provisionary way to access argument calls. Becaause of the current
  instruction set should only allow double precision numbers, since for strings
  the only possible functionality would be to echo them.
  This is to be done through an instruction not as general as "arg" since that
  should be reserved for a more general way to handle command line arguments.
  The instructions should be `rad` which stands for "Retrieve Argument as Double"
  where the first argument is a register holding the index of the argument and
  the second argument is the destination register.
  Also the general instruction `rac` which stands for "Retrieve Argument Count"
  will be implemented such that it's first argument is the register to receive
  the argument count.
  This depends on PC018.
PC020:  2016-09-17
  Make it so the exit from getLiteral and getRegister in compiler.go become
  actual compiler errors that report file name and line number.
PC021:  2016-09-17
  Remove all panic calls.
PC023:  2016-09-17
  Add other jump instructions.
  jge idx     Jump if greater or equal
  jle idx     Jump if less or equal
  jne idx     Jump if not equal
PC025:	2016-09-17
  Adds an instruction for swapping.
  swp reg reg Swaps the values of a and b
PC027:	2016-09-17
  Implements instructions for retrieving date and time.
  dty	reg		reg := Current year
  dtm	reg		reg := Current month
  dtd	reg		reg := Current day
  tmh	reg		reg := Current hour
  tmm	reg		reg := Current minute
  tms	reg		reg := Current seconds
  now	reg		reg := Number of nanoseconds elapsed since January 1, 1970 UTC.
PC028:	2016-09-17
  Implements instructions for random number generation.
  rns	reg		- Sets the random number generation seed to reg
  rnd	reg		- Sets reg to a random number uniformily distributed within (0,1)
  rnn	reg		- Sets reg to a random number extracted from a standard normal

  2. Discarded

PC007:	Substituted by PC018
  Change parsing into compilation of a bytecode binary file.
PC009:	Not applicable
  Not a language change but change the convention for function calls so that the
  first four values passed to a function to be in the %0, %1, %2 and %3 registers
  respectively.

	3. Proposed

PC011:
  Having stack frame manipulation in mind the instructions adl and sul should be
  added in order to facilitate allocation of compile time sizes.
  adl	lit	reg		reg := reg + lit
  sul	lit	reg		reg := reg - lit
PC012:
  Implements general stack based memory access.
  A new memory stack is implemented.

run.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

func runtimeError(msg string, code, index float64) {
	exit(fmt.Sprintf("runtime: %s [%0.f@%0.f]", msg, code, index))
}

func run(code []float64) {
	count := 0
	reader := bufio.NewReader(os.Stdin)

	flt := false
	feq := false
	fgt := false

	var reg [regCount]float64

	var callStack [callStackSize]int
	csp := 0

	var stack [stackSize]float64
	sp := 0

	halted := false

	for {
		if count >= len(code) || halted {
			break
		}

		// Debug
		// fmt.Println(b2s[code[count]])

		switch code[count] {
		case hlt:
			halted = true
		case nop:
			count += 1

		case cll:
			callStack[csp] = count + 2
			csp += 1
			count = int(code[count+1])
		case ret:
			csp -= 1
			count = callStack[csp]

		case val:
			reg[int(code[count+2])] = code[count+1]
			count += 3
		case mov:
			reg[int(code[count+2])] = reg[int(code[count+1])]
			count += 3
		case swp:
			tmp := reg[int(code[count+2])]
			reg[int(code[count+2])] = reg[int(code[count+1])]
			reg[int(code[count+1])] = tmp
			count += 3
		case psh:
			stack[sp] = reg[int(code[count+1])]
			sp += 1
			count += 2
		case pop:
			sp -= 1
			reg[int(code[count+1])] = stack[sp]
			count += 2

		case rac:
			reg[int(code[count+1])] = float64(len(os.Args)) - 2
			count += 2
		case rad:
			val, err := strconv.ParseFloat(os.Args[int(reg[int(code[count+1])])+2], 64)
			if err != nil {
				runtimeError("invalid number", rad, float64(count))
				break
			}
			reg[int(code[count+2])] = val
			count += 3

		case add:
			reg[int(code[count+2])] += reg[int(code[count+1])]
			count += 3
		case sub:
			reg[int(code[count+2])] -= reg[int(code[count+1])]
			count += 3
		case mul:
			reg[int(code[count+2])] *= reg[int(code[count+1])]
			count += 3
		case div:
			reg[int(code[count+2])] /= reg[int(code[count+1])]
			count += 3
		case inc:
			reg[int(code[count+1])] += 1
			count += 2
		case dec:
			reg[int(code[count+1])] -= 1
			count += 2

		case flr:
			reg[int(code[count+1])] = math.Floor(reg[int(code[count+1])])
			count += 2
		case cel:
			reg[int(code[count+1])] = math.Ceil(reg[int(code[count+1])])
			count += 2
		case abs:
			reg[int(code[count+1])] = math.Abs(reg[int(code[count+1])])
			count += 2
		case sqr:
			reg[int(code[count+1])] = math.Sqrt(reg[int(code[count+1])])
			count += 2
		case sin:
			reg[int(code[count+1])] = math.Sin(reg[int(code[count+1])])
			count += 2
		case cos:
			reg[int(code[count+1])] = math.Cos(reg[int(code[count+1])])
			count += 2
		case tan:
			reg[int(code[count+1])] = math.Tan(reg[int(code[count+1])])
			count += 2
		case exp:
			reg[int(code[count+1])] = math.Exp(reg[int(code[count+1])])
			count += 2
		case log:
			reg[int(code[count+1])] = math.Log(reg[int(code[count+1])])
			count += 2
		case gam:
			reg[int(code[count+1])] = math.Gamma(reg[int(code[count+1])])
			count += 2
		case pow:
			reg[int(code[count+2])] = math.Pow(reg[int(code[count+2])], reg[int(code[count+1])])
			count += 2
		case pwa:
			reg[int(code[count+2])] = math.Pow(reg[int(code[count+1])], reg[int(code[count+2])])
			count += 2

		case jmp:
			count = int(code[count+1])
		case jlt:
			if flt {
				count = int(code[count+1])
			} else {
				count += 2
			}
		case jeq:
			if feq {
				count = int(code[count+1])
			} else {
				count += 2
			}
		case jgt:
			if fgt {
				count = int(code[count+1])
			} else {
				count += 2
			}
		case jge:
			if !flt {
				count = int(code[count+1])
			} else {
				count += 2
			}
		case jle:
			if !fgt {
				count = int(code[count+1])
			} else {
				count += 2
			}
		case jne:
			if !feq {
				count = int(code[count+1])
			} else {
				count += 2
			}

		case cmp:
			if reg[int(code[count+1])] < reg[int(code[count+2])] {
				flt = true
				feq = false
				fgt = false
			} else if reg[int(code[count+1])] == reg[int(code[count+2])] {
				flt = false
				feq = true
				fgt = false
			} else {
				flt = false
				feq = false
				fgt = true
			}
			count += 3
		case cmz:
			if reg[int(code[count+1])] < 0 {
				flt = true
				feq = false
				fgt = false
			} else if reg[int(code[count+1])] == 0 {
				flt = false
				feq = true
				fgt = false
			} else {
				flt = false
				feq = false
				fgt = true
			}
			count += 2

		case shw:
			fmt.Println(reg[int(code[count+1])])
			count += 2
		case get:
			txt, err := reader.ReadString('\n')
			if err != nil {
				runtimeError("could not read the input string", get, float64(count))
				break
			}
			val, err := strconv.ParseFloat(strings.TrimSpace(txt), 64)
			if err != nil {
				runtimeError("invalid number", get, float64(count))
				break
			}
			reg[int(code[count+1])] = val
			count += 2
		case dsp:
			fmt.Println(data[int(code[count+1])])
			count += 2

		case dty:
			reg[int(code[count+1])] = float64(time.Now().Year())
			count += 2
		case dtm:
			reg[int(code[count+1])] = float64(time.Now().Month())
			count += 2
		case dtd:
			reg[int(code[count+1])] = float64(time.Now().Day())
			count += 2
		case tmh:
			reg[int(code[count+1])] = float64(time.Now().Hour())
			count += 2
		case tmm:
			reg[int(code[count+1])] = float64(time.Now().Minute())
			count += 2
		case tms:
			reg[int(code[count+1])] = float64(time.Now().Second())
			count += 2
		case now:
			reg[int(code[count+1])] = float64(time.Now().UTC().UnixNano())
			count += 2

		case sed:
			rand.Seed(int64(reg[int(code[count+1])]))
			count += 2
		case rnd:
			reg[int(code[count+1])] = rand.Float64()
			count += 2
		case rnn:
			reg[int(code[count+1])] = rand.NormFloat64()
			count += 2

		default:
			runtimeError("invalid instruction", code[count], float64(count))
		}
	}
}


main.go
package main

import (
	"fmt"
	"os"
)

func exit(msg string) {
	fmt.Printf("\033[1m\033[31mgovm\033[39m\033[0m: %s\n", msg)
	os.Exit(1)
}

const (
	regCount      = 16
	stackSize     = 256
	callStackSize = 64
)

func main() {
	if len(os.Args) < 2 {
		exit("incorrect number of arguments")
	}

	compile := false

	if os.Args[1] == "version" {
		fmt.Println("GoVM 0.2")
		os.Exit(0)
	} else if os.Args[1] == "compile" {
		compile = true
	}

	if compile {
		if len(os.Args) < 4 {
			exit("compilation needs at least 2 extra arguments")
		}

		// Make sure all files exist
		for _, arg := range os.Args[3:] {
			if _, err := os.Stat(arg); os.IsNotExist(err) {
				exit(fmt.Sprintf("file \"%s\" does not exist", arg))
			}
		}

		// Main file
		file, err := os.Open(os.Args[3])
		if err != nil {
			exit(fmt.Sprintf("file \"%s\" could not be opened", os.Args[3]))
		}

		code, count := parser(file, os.Args[3], 0)
		file.Close()

		// Extra files
		for _, arg := range os.Args[4:] {
			file, err := os.Open(arg)
			if err != nil {
				exit(fmt.Sprintf("file \"%s\" could not be opened", arg))
			}

			var newCode []float64
			newCode, count = parser(file, arg, count)
			code = append(code, newCode...)
			file.Close()
		}

		// Debug
		// fmt.Println(decode(code))

		evaluateLabels(code)

		writeFile(code, os.Args[2])
	} else {
		code := loadFile(os.Args[1])

		// Debug
		// fmt.Println(decode(code))

		run(code)
	}
}


compiler.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func compilerError(msg, file, str string, line int) {
	exit(fmt.Sprintf("compiler: %s [%s] at line %d in %s", msg, str, line, file))
}

func encoderError(msg string) {
	exit(fmt.Sprintf("encoder: %s", msg))
}

func decoderError(msg string) {
	exit(fmt.Sprintf("decoder: %s", msg))
}

// We use this so that if a compilation error occurr we know the faulty line
type labelInfo struct {
	name string
	file string
	line int
}

var labels map[string]int
var labelsPending map[int]labelInfo

var data []string
var dataCurr int
var dataMap map[string]int

func init() {
	labels = make(map[string]int)
	labelsPending = make(map[int]labelInfo)
	data = make([]string, 0)
	dataCurr = 0
	dataMap = make(map[string]int)
}

func parser(f *os.File, fileName string, start int) (code []float64, count int) {
	code = make([]float64, 0)

	reader := bufio.NewReader(f)

	lineNumber := 0
	count = start

	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			} else {
				compilerError("failed line read", fileName, "ReadString()", lineNumber)
			}
		}
		lineNumber += 1

		ins := strings.Fields(line)

		if len(ins) == 0 {
			continue
		}

		switch ins[0] {
		// Pseudo-instructions

		case "str":
			data = append(data, strings.Join(ins[2:len(ins)], " "))
			dataMap[ins[1]] = dataCurr
			dataCurr += 1

		case "lab":
			if labels[ins[1]] != 0 {
				compilerError("label redefinition", fileName, ins[1], lineNumber)
			}
			labels[ins[1]] = count

		// Instructions

		case "hlt":
			code = append(code, hlt)
			count += 1
		case "nop":
			break

		case "cll":
			code = append(code, cll)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "ret":
			code = append(code, ret)
			count += 1

		case "val":
			code = append(code, val)
			code = append(code, getLiteral(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "mov":
			code = append(code, mov)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "swp":
			code = append(code, swp)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "psh":
			code = append(code, psh)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "pop":
			code = append(code, pop)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2

		case "rac":
			code = append(code, rac)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "rad":
			code = append(code, rad)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3

		case "add":
			code = append(code, add)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "sub":
			code = append(code, sub)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "mul":
			code = append(code, mul)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "div":
			code = append(code, div)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "inc":
			code = append(code, inc)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "dec":
			code = append(code, dec)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2

		case "flr":
			code = append(code, flr)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "cel":
			code = append(code, cel)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "abs":
			code = append(code, abs)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "sqr":
			code = append(code, sqr)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "sin":
			code = append(code, sin)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "cos":
			code = append(code, cos)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "tan":
			code = append(code, tan)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "exp":
			code = append(code, exp)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "log":
			code = append(code, log)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "gam":
			code = append(code, gam)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "pow":
			code = append(code, pow)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "pwa":
			code = append(code, pwa)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3

		case "jmp":
			code = append(code, jmp)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jlt":
			code = append(code, jlt)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jeq":
			code = append(code, jeq)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jgt":
			code = append(code, jgt)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jge":
			code = append(code, jge)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jle":
			code = append(code, jle)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2
		case "jne":
			code = append(code, jne)
			code = append(code, 0.0) // Placeholder
			labelsPending[count+1] = labelInfo{ins[1], fileName, lineNumber}
			count += 2

		case "cmp":
			code = append(code, cmp)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			code = append(code, getRegister(ins[2], fileName, lineNumber))
			count += 3
		case "cmz":
			code = append(code, cmz)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2

		case "shw":
			code = append(code, shw)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "get":
			code = append(code, get)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "dsp":
			code = append(code, dsp)
			code = append(code, float64(dataMap[ins[1]]))
			count += 2

		case "dty":
			code = append(code, dty)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "dtm":
			code = append(code, dtm)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "dtd":
			code = append(code, dtd)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "tmh":
			code = append(code, tmh)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "tmm":
			code = append(code, tmm)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "tms":
			code = append(code, tms)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "now":
			code = append(code, now)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2

		case "sed":
			code = append(code, sed)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "rnd":
			code = append(code, rnd)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2
		case "rnn":
			code = append(code, rnn)
			code = append(code, getRegister(ins[1], fileName, lineNumber))
			count += 2

		default:
			if ins[0][len(ins[0])-1] == ':' {
				labName := ins[0][:len(ins[0])-1]
				if labels[labName] != 0 {
					compilerError("label redefinition", fileName, labName, lineNumber)
				}
				labels[labName] = count
				break
			}

			compilerError("invalid instruction", fileName, ins[0], lineNumber)
		}
	}

	return
}

func evaluateLabels(code []float64) {
	for index, info := range labelsPending {
		code[index] = float64(labels[info.name])
		if code[index] == 0 {
			compilerError("undefined label", info.file, info.name, info.line)
		}
	}
}

func getLiteral(ins, file string, line int) float64 {
	lit, err := strconv.ParseFloat(ins, 64)
	if err != nil {
		compilerError("invalid literal", file, ins, line)
	}
	return lit
}

func getRegister(ins, file string, line int) float64 {
	tins := ins[1:len(ins)]
	reg, err := strconv.ParseFloat(tins, 64)
	if err != nil {
		compilerError("invalid register", file, ins, line)
	}
	return float64(reg)
}

func writeFile(code []float64, fileName string) {
	file, err := os.Create(fileName)
	defer file.Close()
	if err != nil {
		encoderError("could not create binary file")
	}

	// Header
	var headerValue int64
	headerValue = 10002
	err = binary.Write(file, binary.LittleEndian, headerValue)
	if err != nil {
		encoderError("could not write header")
	}

	// Code size
	var codeSize int64
	codeSize = int64(len(code))
	err = binary.Write(file, binary.LittleEndian, codeSize)
	if err != nil {
		encoderError("could not write code size")
	}

	// Code
	for i := 0; i < len(code); i++ {
		err = binary.Write(file, binary.LittleEndian, code[i])
		if err != nil {
			encoderError("could not write code entry")
		}
	}

	// Data size
	var dataStackSize int64
	dataStackSize = int64(dataCurr)
	err = binary.Write(file, binary.LittleEndian, dataStackSize)
	if err != nil {
		encoderError("could not write data size")
	}

	// Data
	for i := 0; i < dataCurr; i++ {
		_, err = fmt.Fprintln(file, data[i])
		if err != nil {
			encoderError("could not write data entry")
		}
	}
}

func loadFile(fileName string) (code []float64) {
	file, err := os.Open(fileName)
	defer file.Close()
	if err != nil {
		decoderError("could not open file")
	}

	var headerValue int64
	err = binary.Read(file, binary.LittleEndian, &headerValue)
	if err != nil {
		decoderError("could not read header")
	}
	if headerValue != 10002 {
		decoderError("invalid header value")
	}

	var codeSize int64
	err = binary.Read(file, binary.LittleEndian, &codeSize)
	if err != nil {
		decoderError("could not read code size")
	}
	code = make([]float64, codeSize)
	for i := 0; i < int(codeSize); i++ {
		err = binary.Read(file, binary.LittleEndian, &code[i])
		if err != nil {
			decoderError("could not read code entry")
		}
	}

	var dataStackSize int64
	err = binary.Read(file, binary.LittleEndian, &dataStackSize)
	if err != nil {
		decoderError("could not read data size")
	}

	r := bufio.NewReader(file)
	for i := 0; i < int(dataStackSize); i++ {
		str, err := r.ReadString('\n')
		if err != nil {
			decoderError("could not read data entry")
		}
		data = append(data, str[:len(str)-1])
	}

	return
}
